golang/x/relay/keeper/querier.go

```diff
@@ -58,24 +89,243 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
+	if err != nil {
+		return []byte{}, types.ErrBadHex(types.DefaultCodespace)
+	}
+	currentBestDigest, newDigestErr := btcspv.NewHash256Digest(currentBestBytes)
+	if newDigestErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
+	}
+
+	newBest := path[2]
+	newBestBytes, decodeErr := hex.DecodeString(newBest)
+	if err != nil {
+		return []byte{}, types.ErrBadHex(types.DefaultCodespace)
+	}
+	newBestDigest, newDigestErr := btcspv.NewHash256Digest(newBestBytes)
+	if newDigestErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
+	}
+
+	limit, convErr := strconv.Atoi(path[3])
+	if convErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, convErr)
+	}
+	newLimit := uint32(limit)
+
+	// This calls the keeper with the parsed arguments, and gets an answer
+	result, err := keeper.HeaviestFromAncestor(ctx, ancestor, currentBestDigest, newBestDigest, newLimit)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	// Now we format the answer as a response
+	response := types.QueryResHeaviestFromAncestor{
+		Ancestor:    ancestor,
+		CurrentBest: currentBestDigest,
+		NewBest:     newBestDigest,
+		Limit:       newLimit,
+		Res:         result,
+	}
+
+	// And we serialize that response as JSON
+	res, marshalErr := codec.MarshalJSONIndent(keeper.cdc, response)
+	if marshalErr != nil {
+		return []byte{}, types.ErrMarshalJSON(types.DefaultCodespace)
+	}
+	return res, nil
+}
+
+func queryIsMostRecentCommonAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keeper Keeper) (res []byte, err sdk.Error) {
+	// check that the path is this many items long, error if not
+	if len(path) > 4 {
+		return []byte{}, types.ErrTooManyArguments(types.DefaultCodespace)
+	} else if len(path) < 4 {
+		return []byte{}, types.ErrNotEnoughArguments(types.DefaultCodespace)
+	}
+
+	ancestorHex := path[0]
+	ancestorBytes, decodeErr := hex.DecodeString(ancestorHex)
+	if decodeErr != nil {
+		return []byte{}, types.ErrBadHex(types.DefaultCodespace)
+	}
+	ancestor, newDigestErr := btcspv.NewHash256Digest(ancestorBytes)
+	if newDigestErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
+	}
+
+	// parse the second path item as hex.
+	left := path[1]
+	leftBytes, decodeErr := hex.DecodeString(left)
+	if err != nil {
+		return []byte{}, types.ErrBadHex(types.DefaultCodespace)
+	}
+	leftDigest, newDigestErr := btcspv.NewHash256Digest(leftBytes)
+	if newDigestErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
+	}
+
+	right := path[2]
+	rightBytes, decodeErr := hex.DecodeString(right)
+	if err != nil {
+		return []byte{}, types.ErrBadHex(types.DefaultCodespace)
+	}
+	rightDigest, newDigestErr := btcspv.NewHash256Digest(rightBytes)
+	if newDigestErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
+	}
+
+	limit, convErr := strconv.Atoi(path[3])
```

Comment:
use `ParseUint` instead

also, if path is only 3 long, use a default number, as above

Reply:
Does it matter what default number? I think you mentioned 240 before...

Reply:
But earlier we used 15